source/probability_theory_4_finite.Rmd

```diff
@@ -524,284 +481,178 @@ sampling mechanism as in Example 8-1, but now we record "yes" for a
+  or more pairs match, record "yes," otherwise "no."
 
-**Step 4.** Repeat steps (2) and (3) perhaps 200 times.
+**Step 4.** Repeat steps (2) and (3), say, 10,000 times.
 
 **Step 5.** Count the proportion "yes." This estimates the probability
 of 4 or more pairs.
 
 Exercise for the student: Write the steps to do this example with random
-numbers.
-
-Now let's examine the RESAMPLING STATS solution "Couples." The key step
-is to fill array A with 10 elements numbered "1" to "10" in order, with
-each number standing for a male. We fill array B in similar fashion to
-represent the females.
-
-In each of the 1000 samples we SHUFFLE the elements of B, which stands
-for the females, to create a new vector C. As long as *either* the males
-or the females are randomly ordered, the probability of getting a
-correct match is determined by chance. There are several ways to
-determine whether a match occurs. Our method is using the SUBTRACT
-command to compare each element of A with each element of C. A match
-causes a "0," this would mean that the third couple was matched
-correctly because they have the same numbers---"3"s. We then COUNT the
-number of "0"s in D to determine the number of couples correctly matched
-in the current sample. The result for the current sample is placed in E,
-then transferred to Z in order to keep SCORE for each of the 1000
-samples. This ends the loop.
-
-`           NUMBERS 1,10 a         `
-
-An array of 10 males.
-
-`           NUMBERS 1,10 b         `
-
-An identical array of 10 females---the pair for each of the males.
-
-`           REPEAT 1000         `
-
-Do the experiment 1000 times.
+numbers.  The solution follows below.
 
-`           SHUFFLE b c         `
-
-Shuffle the females.
-
-`           END         `
-
-`           SUBTRACT a c d         `
-
-This operation pairs each shuffled female with a male and subtracts. If
-it is an original pairing (1/1, 2/2, etc.), the result will be a 0. The
-number of 0's indicates how many of the 10 males got paired up again
-with their original partner.
-
-`           COUNT d =0 e         `
-
-Count the number of 0's and put the result in e.
-
-`           SCORE e z         `
-
-Keep track of each trial result.
-
-End the trial, go back and repeat until all 1000 are complete.
-
-`           HISTOGRAM z         `
-
-Produce a histogram of the results.
+```{python, opts.label="py_ed"}
+N = 10000
+trial_results = np.zeros(N, dtype=int)
 
-![](images/12-Chap-8_007.png)
+# Assign numbers to each student, according to their pair, after the first
+# icebreaker
+birmingham = np.arange(10)
+berkeley = np.arange(10)
 
-From this histogram, we see that in about 3 percent of the trials did 4
-or more couples end up being re-paired with their own partners. We can
-calculate this proportion directly with RESAMPLING STATS:
+for i in range(N):
+    # Randomly shuffle the students from berkeley
+    # (We could also have shuffled both arrays---it does not alter the outcome)
```

Comment:
May be better to just shuffle both here, and say that, in fact, we don't need to shuffle the second set - it iisn't obvious to students why you don't need to shuffle both.